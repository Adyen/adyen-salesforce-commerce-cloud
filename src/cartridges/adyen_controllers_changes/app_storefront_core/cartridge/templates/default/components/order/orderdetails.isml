<iscontent type="text/html" charset="UTF-8" compact="true"/>

<iscomment>
	Displays order details, such as order number, creation date, payment information,
	order totals and information for each contained shipment.
	This template module can be used in order confirmation pages as well as in the
	order history to render the details of a given order. Depending on the context
	being used in, one might omit rendering certain information.

	Parameters:

	order		: the order whose details to render
	orderstatus	: if set to true, the order status will be rendered
				  if set to false or not existing, the order status will not be rendered
</iscomment>

<isset name="Order" value="${pdict.order}" scope="page"/>

<div class="orderdetails">
	<div class="order-information">
		<h1 class="order-number">
			<span class="label"><i class="fa fa-cube"></i>${Resource.msg('order.orderdetails.ordernumber','order',null)}</span>
			<span class="value"><isprint value="${Order.orderNo}"/></span>
		</h1>
		<iscomment>Order Status</iscomment>
		<isif condition="${!empty(pdict.orderstatus) && pdict.orderstatus == 'true'}">
			<isinclude template="account/orderhistory/orderstatusinclude"/>
		</isif>
		<div class="order-date">
			<span class="label">${Resource.msg('order.orderdetails.orderplaced','order',null)}</span>
			<span class="value"><isprint value="${Order.creationDate}" style="DATE_LONG"/></span>
		</div>
	</div>
	<div class="order-payment-summary">
		<div class="label">${Resource.msg('order.orderdetails.paymenttotal','order',null)}</div>
		<div class="order-detail-summary">
			<isif condition="${Order.shipments.length > 1}">
				<isordertotals p_lineitemctnr="${Order}" p_showshipmentinfo="${true}" p_shipmenteditable="${false}" p_totallabel="${Resource.msg('global.ordertotal','locale',null)}"/>
			<iselse/>
				<isordertotals p_lineitemctnr="${Order}" p_showshipmentinfo="${false}" p_shipmenteditable="${false}" p_totallabel="${Resource.msg('global.ordertotal','locale',null)}"/>
			</isif>
		</div>
	</div>
	<div class="order-payment-instruments">
		<div class="label">
			<isif condition="${Order.paymentInstruments.length == 1}">
				${Resource.msg('order.orderdetails.paymentmethod','order',null)}
			<iselse/>
				${Resource.msg('order.orderdetails.paymentmethods','order',null)}
			</isif>
		</div>
		<isloop items="${Order.getPaymentInstruments()}" var="paymentInstr">
			<isif condition="${dw.order.PaymentInstrument.METHOD_GIFT_CERTIFICATE.equals(paymentInstr.paymentMethod)}">
				<div class="orderpaymentinstrumentsgc">
					<span class="label"><isprint value="${dw.order.PaymentMgr.getPaymentMethod(paymentInstr.paymentMethod).name}" /></span>
					<span class="value"><isprint value="${paymentInstr.maskedGiftCertificateCode}"/></span>
					<span class="payment-amount">
						<span class="label">${Resource.msg('global.amount','locale',null)}:</span>
						<span class="value"><isprint value="${paymentInstr.paymentTransaction.amount}"/></span>
					</span>
				</div>
			<iselse/>
				<div class="payment-type"><isprint value="${dw.order.PaymentMgr.getPaymentMethod(paymentInstr.paymentMethod).name}" /></div>
				<isminicreditcard card="${paymentInstr}" show_expiration="${false}"/>
				<div class="payment-amount">
					<span class="label">${Resource.msg('global.amount','locale',null)}:</span>
					<span class="value"><isprint value="${paymentInstr.paymentTransaction.amount}"/></span>
				</div><!-- END: payment-amount -->
			</isif>
		</isloop>
	</div>
	<div class="order-billing">
		<div class="label">${Resource.msg('order.orderdetails.billingaddress','order',null)}</div>
		<isminiaddress p_address="${Order.billingAddress}"/>
	</div>

	<iscomment>render a box for each shipment</iscomment>
	<div class="order-shipments">
		<isloop items="${Order.shipments}" var="shipment" status="shipmentloopstate">
			<h2>${Resource.msgf('multishippingshipments.shipment','checkout',null, shipmentloopstate.count)}</h2>

			<isif condition="${shipment.productLineItems.size() > 0}">
				<iscomment>Shipment Items</iscomment>
				<div class="order-shipment-table">
					<div class="order-shipment-details">
						<div class="order-shipment-address">
							<isif condition="${!empty(shipment.custom.shipmentType) && shipment.custom.shipmentType == 'instore'}">
								<div class="label">${Resource.msg('cart.store.instorepickup','checkout',null)}</div>
							<iselse/>
								<div class="label">${Resource.msg('order.orderdetails.shippingto','order',null)}</div>
							</isif>
							<isminishippingaddress p_shipment="${shipment}" p_editable="${false}" p_showmethod="${false}" p_showpromos="${false}"/>
						</div>
						<isif condition="${shipment.custom.shipmentType != 'instore'}">
							<div class="shipping-status">
								<div class="label">${Resource.msg('order.orderdetails.shippingstatus','order',null)}</div>
								<isif condition="${shipment.shippingStatus==dw.order.Shipment.SHIPPING_STATUS_NOTSHIPPED}">
									<div class="value">${Resource.msg('order.orderdetails.notshipped','order',null)}</div>
								<iselseif condition="${shipment.shippingStatus==dw.order.Shipment.SHIPPING_STATUS_SHIPPED}">
									<div class="value">${Resource.msg('order.orderdetails.shipped','order',null)}</div>
								<iselse/>
									<div class="value">${Resource.msg('order.orderdetails.notknown','order',null)}</div>
								</isif>
							</div>
						</isif>
						<isif condition="${!empty(shipment.trackingNumber)}">
							<div class="trackingnumber">
								<div class="label">${Resource.msg('order.orderdetails.tracking','order',null)}</div>
								<div class="value"><!-- Tracking Number --><isprint value="${shipment.trackingNumber}"/></div>
							</div>
						</isif>
						<div class="shipping-method">
							<div class="label">${Resource.msg('order.orderdetails.shippingmethod','order',null)}</div>
							<isif condition="${!empty(shipment.shippingMethod)}">
								<div class="value"><isprint value="${shipment.shippingMethod.displayName}"/></div>
							<iselse/>
								<div class="value"><isprint value="${shipment.shippingMethodID}"/></div>
							</isif>
						</div>
						<iscomment>Shipment Gift Message</iscomment>
						<isif condition="${shipment.gift}">
							<isif condition="${!empty(shipment.giftMessage)}">
								<div class="order-shipment-gift-message">
									<div class="label">${Resource.msg('order.orderdetails.giftmessage','order',null)}</div>
									<div class="value"><isprint value="${shipment.giftMessage}"/></div>
								</div>
							</isif>
						</isif>
					</div>
					<div class="line-items">
						<isloop items="${shipment.productLineItems}" var="productLineItem" status="pliloopstate">
							<div class="line-item">
								<div class="line-item-details">
									<isif condition="${pliloopstate.first}">
										<div class="label">${Resource.msg('global.item','locale',null)}</div>
									</isif>
									<iscomment>Display product line and product using module</iscomment>
									<isdisplayliproduct p_productli="${productLineItem}" p_editable="${false}"/>
								</div>
								<div class="line-item-quantity">
									<isif condition="${pliloopstate.first}">
										<div class="label">${Resource.msg('global.qty','locale',null)}</div>
									</isif>
									<isprint value="${productLineItem.quantity}"/>
								</div>
								<div class="line-item-price">
									<isif condition="${pliloopstate.first}">
										<div class="label">${Resource.msg('global.price','locale',null)}</div>
									</isif>
									<iscomment>Render quantity. If it is a bonus product render word 'Bonus'</iscomment>
									<isif condition="${productLineItem.bonusProductLineItem}">
										<div class="bonus-item">
											<isset name="bonusProductPrice" value="${productLineItem.getAdjustedPrice()}" scope="page"/>
											<isinclude template="checkout/components/displaybonusproductprice" />
											<isprint value="${bonusProductPriceValue}" />
										</div>
									<iselse/>
										<isprint value="${productLineItem.adjustedPrice}"/>
										<isif condition="${productLineItem.optionProductLineItems.size() > 0}">
											<isloop items="${productLineItem.optionProductLineItems}" var="optionLI">
												<div class="product-option-price">+ <isprint value="${optionLI.adjustedPrice}"/></div>
											</isloop>
										</isif>
									</isif>
								</div>
							</div>
						</isloop>
					</div>
				</div>
			</isif>

			<isif condition="${shipment.giftCertificateLineItems.size() > 0}">
				<div class="order-shipment-table">
					<div class="order-shipment-details">
						<iscomment>Shipment Gift Certificate</iscomment>
						<isloop items="${shipment.giftCertificateLineItems}" var="giftCertificateLineItem" status="gcliloopstate">
							<div class="order-gift-cert-attributes order-shipment-address">
								<div class="order-gift-cert-to">
									<div class="label">${Resource.msg('order.orderdetails.giftcertto','order',null)}</div>
									<div class="value"><isprint value="${giftCertificateLineItem.recipientName}"/></div>
									<div class="value"><isprint value="${giftCertificateLineItem.recipientEmail}"/></div>
								</div>
								<div class="order-gift-cert-from">
									<div class="label">${Resource.msg('order.orderdetails.giftcertfrom','order',null)}</div>
									<div class="value"><isprint value="${giftCertificateLineItem.senderName}"/></div>
									<div class="value"><isprint value="${Order.customerEmail}"/></div>
								</div>
							</div>
							<div class="order-gift-cert-amount">
								<div class="label">${Resource.msg('global.giftcertificate','locale',null)}</div>
								<div class="value"><isprint value="${giftCertificateLineItem.price}"/></div>
							</div>
							<div class="shipping-method">
								<div class="label">${Resource.msg('order.orderdetails.shippingmethod','order',null)}</div>
								<div class="value">${Resource.msg('order.orderdetails.giftcertshipping','order',null)}</div>
								<div class="value"><isprint value="${giftCertificateLineItem.recipientEmail}"/></div>
							</div>
						</isloop>

						<iscomment>if shipment is marked as gift</iscomment>
						<isif condition="${shipment.gift}">
							<div class="order-shipment-gift-message">
								<div class="label">${Resource.msg('order.orderdetails.giftmessage','order',null)}</div>
								<isif condition="${!empty(shipment.giftMessage)}">
									<div class="value"><isprint value="${shipment.giftMessage}"/></div>
								</isif>
							</div>
						<iselse/>
							<div class="order-shipment-gift-message">
								<isset name="theGiftCert" value="${shipment.giftCertificateLineItems.iterator().next()}" scope="page"/>
								<div class="label">${Resource.msg('order.orderdetails.giftmessage','order',null)}</div>
								<isif condition="${!empty(theGiftCert.message)}">
									<div class="value"><isprint value="${theGiftCert.message}"/></div>
								</isif>
							</div>
						</isif>
					</div>
				</div>
			</isif>
		</isloop>
	</div>
</div>
