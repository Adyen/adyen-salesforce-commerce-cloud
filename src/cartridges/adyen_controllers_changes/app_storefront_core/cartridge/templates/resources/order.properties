##############################################
# Resource bundle: order
##############################################

##############################################
# Template name: order/orderconfirmation-email
##############################################
order.orderconfirmation-email.jumptostore=Go to Our Store
order.orderconfirmation-email.001=Confirmation of Your Order
order.orderconfirmation-email.confirmnumber=Order Confirmation Number:
order.orderconfirmation-email.thankyou=- Thank you for your order!
order.orderconfirmation-email.storeaddress=5 Wall Street
order.orderconfirmation-email.storelocation=Burlington, MA 01803
order.orderconfirmation-email.storephone=Phone: +1.800.555.0199
order.orderconfirmation-email.print=Please print and save a copy for reference.
order.orderconfirmation-email.thankyouus=Thank you for shopping with us!
order.orderconfirmation-email.pleasenote=PLEASE NOTE:
order.orderconfirmation-email.note=Since we begin processing your order shortly after you submit it on our web site, if any changes are necessary you MUST contact us via PHONE ONLY, at 
order.orderconfirmation-email.phone=+1.800.555.0199
order.orderconfirmation-email.shipmentnumber=Shipment#

##############################################
# ISML Directory: components/order/
##############################################
order.orderdetails.paymentinfo=Payment Information
order.orderdetails.billingaddress=Billing Address
order.orderdetails.paymentmethod=Payment Method
order.orderdetails.paymentmethods=Payment Methods
order.orderdetails.paymenttotal=Payment Total
order.orderdetails.shippingmethod=Method:
order.orderdetails.shippingstatus=Shipping Status:
order.orderdetails.shippingto=Shipping To
order.orderdetails.notshipped=Not Shipped
order.orderdetails.shipped=Shipped
order.orderdetails.notknown=Not Known
order.orderdetails.tracking=Tracking Number:
order.orderdetails.orderplaced=Order Placed:
order.orderdetails.ordernumber=Order Number:
order.orderdetails.giftcertto=To:
order.orderdetails.giftcertfrom=From:
order.orderdetails.giftmessage=Gift Message:
order.orderdetails.giftcertshippingaddress=Sent To:
order.orderdetails.giftcertshipping=Delivery Via Email

order.ordersummary.ordersubtotal=Order Subtotal
order.ordersummary.subtotal=Subtotal

order.ordersummary.orderdiscount=Order Discount
order.ordersummary.ordershipping=Shipping
order.ordersummary.ordershipping.edit=Edit
order.ordersummary.ordershippingdiscount=Shipping Discount
order.ordersummary.ordertax=Sales Tax