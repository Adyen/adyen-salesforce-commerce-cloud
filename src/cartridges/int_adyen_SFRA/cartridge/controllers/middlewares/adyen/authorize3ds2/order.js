const Transaction = require('dw/system/Transaction');
const URLUtils = require('dw/web/URLUtils');
const COHelpers = require('*/cartridge/scripts/checkout/checkoutHelpers');
const AdyenHelper = require('*/cartridge/scripts/util/adyenHelper');
const { clearForms } = require('../../../utils/index');
const { handlePlaceOrderError } = require('./errorHandler');

function handleOrderConfirm(
  paymentInstrument,
  order,
  result,
  { req, res, next },
) {
  Transaction.begin();
  AdyenHelper.savePaymentDetails(paymentInstrument, order, result);
  order.setPaymentStatus(dw.order.Order.PAYMENT_STATUS_PAID);
  order.setExportStatus(dw.order.Order.EXPORT_STATUS_READY);
  Transaction.commit();
  COHelpers.sendConfirmationEmail(order, req.locale.id);
  clearForms.clearForms();

  // determines SFRA version for backwards compatibility
  if (AdyenHelper.getAdyenSFRA6Compatibility() === true) {
    res.render('orderConfirmForm', {
      orderID: order.orderNo,
      orderToken: order.orderToken,
    });
  } else {
    res.redirect(
      URLUtils.url(
        'Order-Confirm',
        'ID',
        order.orderNo,
        'token',
        order.orderToken,
      ).toString(),
    );
  }
  return next();
}

function handlePlaceOrder(paymentInstrument, order, result, options) {
  // custom fraudDetection
  const fraudDetectionStatus = { status: 'success' };

  // Places the order
  const placeOrderResult = COHelpers.placeOrder(order, fraudDetectionStatus);
  if (placeOrderResult.error) {
    return handlePlaceOrderError(order, options);
  }

  return handleOrderConfirm(paymentInstrument, order, result, options);
}

module.exports = handlePlaceOrder;
