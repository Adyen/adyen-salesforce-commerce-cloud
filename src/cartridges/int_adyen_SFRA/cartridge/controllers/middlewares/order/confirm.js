const OrderMgr = require('dw/order/OrderMgr');
const AdyenHelper = require('*/cartridge/scripts/util/adyenHelper');
const AdyenConfigs = require('*/cartridge/scripts/util/adyenConfigs');

// order-confirm is POST in SFRA v6.0.0. orderID and orderToken are contained in form.
// This was a GET call with a querystring containing ID & token in earlier versions.
function getOrderId(req) {
  return req.form && req.form.orderID ? req.form.orderID : req.querystring.ID;
}

function getOrderToken(req) {
  return req.form && req.form.orderToken
    ? req.form.orderToken
    : req.querystring.token;
}

function handleAdyenGiving(req, res, order) {
  const clientKey = AdyenConfigs.getAdyenClientKey();
  const environment = AdyenHelper.getCheckoutEnvironment();
  const configuredAmounts = AdyenHelper.getDonationAmounts();
  const charityName = AdyenConfigs.getAdyenGivingCharityName();
  const charityWebsite = AdyenConfigs.getAdyenGivingCharityWebsite();
  const charityDescription = AdyenConfigs.getAdyenGivingCharityDescription();
  const adyenGivingBackgroundUrl = AdyenConfigs.getAdyenGivingBackgroundUrl();
  const adyenGivingLogoUrl = AdyenConfigs.getAdyenGivingLogoUrl();

  const paymentInstrument = order.getPaymentInstruments(
    AdyenHelper.getOrderMainPaymentInstrumentType(order),
  )[0];

  const donationAmounts = {
    currency: session.currency.currencyCode,
    values: configuredAmounts,
  };
  const viewData = res.getViewData();
  viewData.adyen = {
    clientKey,
    environment,
    adyenGivingAvailable: true,
    pspReference:
      paymentInstrument.paymentTransaction.custom.Adyen_pspReference,
    donationAmounts: JSON.stringify(donationAmounts),
    charityName,
    charityDescription,
    charityWebsite,
    adyenGivingBackgroundUrl,
    adyenGivingLogoUrl,
  };
  res.setViewData(viewData);
}

function confirm(req, res, next) {
  const orderId = getOrderId(req);
  const orderToken = getOrderToken(req);
  if (orderId && orderToken) {
    const order = OrderMgr.getOrder(orderId, orderToken);
    const paymentInstrument = order.getPaymentInstruments(
      AdyenHelper.getOrderMainPaymentInstrumentType(order),
    )[0];
    const paymentMethod =
      paymentInstrument.paymentTransaction.custom.Adyen_paymentMethod;

    if (
      AdyenHelper.getAdyenGivingConfig(order) &&
      AdyenHelper.isAdyenGivingAvailable(paymentMethod)
    ) {
      handleAdyenGiving(req, res, order);
    }
  }
  return next();
}

module.exports = confirm;
