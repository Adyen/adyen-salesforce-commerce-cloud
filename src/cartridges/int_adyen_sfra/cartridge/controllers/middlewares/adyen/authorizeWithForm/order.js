const Transaction = require('dw/system/Transaction');
const URLUtils = require('dw/web/URLUtils');
const AdyenHelper = require('*/cartridge/scripts/util/adyenHelper');
const COHelpers = require('*/cartridge/scripts/checkout/checkoutHelpers');
const { clearForms } = require('../../../utils/index');

function handleOrderConfirmation(
  paymentInstrument,
  result,
  order,
  { req, res, next },
) {
  Transaction.begin();
  AdyenHelper.savePaymentDetails(paymentInstrument, order, result);
  order.setPaymentStatus(dw.order.Order.PAYMENT_STATUS_PAID);
  order.setExportStatus(dw.order.Order.EXPORT_STATUS_READY);
  Transaction.commit();
  COHelpers.sendConfirmationEmail(order, req.locale.id);
  clearForms.clearForms();

  // determines SFRA version for backwards compatibility
  if (AdyenHelper.getAdyenSFRA6Compatibility() === true) {
    res.render('orderConfirmForm', {
      orderID: order.orderNo,
      orderToken: order.orderToken,
    });
  } else {
    res.redirect(
      URLUtils.url(
        'Order-Confirm',
        'ID',
        order.orderNo,
        'token',
        order.orderToken,
      ).toString(),
    );
  }
  return next();
}

module.exports = handleOrderConfirmation;
