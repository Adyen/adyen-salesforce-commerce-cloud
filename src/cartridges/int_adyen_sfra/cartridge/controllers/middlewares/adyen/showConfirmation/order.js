const Transaction = require('dw/system/Transaction');
const URLUtils = require('dw/web/URLUtils');
const AdyenHelper = require('*/cartridge/scripts/util/adyenHelper');
const { clearForms } = require('../../../utils/index');

function handleOrderConfirm(
  order,
  orderModel,
  adyenPaymentInstrument,
  result,
  { res, next },
) {
  Transaction.wrap(() => {
    order.custom.Adyen_CustomerEmail = JSON.stringify(orderModel);
    AdyenHelper.savePaymentDetails(adyenPaymentInstrument, order, result);
  });

  clearForms.clearForms();
  // determines SFRA version for backwards compatibility
  if (AdyenHelper.getAdyenSFRA6Compatibility() === true) {
    res.render('orderConfirmForm', {
      orderID: order.orderNo,
      orderToken: order.orderToken,
    });
  } else {
    res.redirect(
      URLUtils.url(
        'Order-Confirm',
        'ID',
        order.orderNo,
        'token',
        order.orderToken,
      ).toString(),
    );
  }
  return next();
}

module.exports = handleOrderConfirm;
