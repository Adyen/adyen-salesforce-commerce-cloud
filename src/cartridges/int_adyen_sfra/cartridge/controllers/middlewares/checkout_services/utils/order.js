const Resource = require('dw/web/Resource');
const URLUtils = require('dw/web/URLUtils');
const COHelpers = require('*/cartridge/scripts/checkout/checkoutHelpers');
const hooksHelper = require('*/cartridge/scripts/helpers/hooks');
const addressHelpers = require('*/cartridge/scripts/helpers/addressHelpers');
const { fraudDetection } = require('*/cartridge/scripts/hooks/fraudDetection');
const { hasAdyenPaymentMethod } = require('../helpers/index');
const handleTransaction = require('./transaction');
const handlePaymentAuthorization = require('./payment');
const handleFraudDetection = require('./fraud');

function createOrder(currentBasket, { res, req, next }, emit) {
  const validateOrder = (order) => {
    // Creates a new order.
    if (!order) {
      res.json({
        error: true,
        errorMessage: Resource.msg('error.technical', 'checkout', null),
      });
      emit('route:Complete');
    }
    return !!order;
  };

  const handlePlaceOrder = (order, fraudDetectionStatus) => {
    const placeOrderResult = COHelpers.placeOrder(order, fraudDetectionStatus);
    if (placeOrderResult.error) {
      res.json({
        error: true,
        errorMessage: Resource.msg('error.technical', 'checkout', null),
      });
      emit('route:Complete');
    }

    return !placeOrderResult.error;
  };

  const validateOrderAndAuthorize = (order) => {
    const isValidOrder = validateOrder(order);
    if (isValidOrder) {
      const isAuthorized = handlePaymentAuthorization(
        order,
        { req, res },
        emit,
      );
      return isAuthorized;
    }
    return false;
  };

  const handleCreateOrder = (order) => {
    const isAuthorized = validateOrderAndAuthorize(order);
    if (isAuthorized) {
      const fraudDetectionStatus = hooksHelper(
        'app.fraud.detection',
        'fraudDetection',
        currentBasket,
        fraudDetection,
      );
      const isSuccessful = handleFraudDetection(
        fraudDetectionStatus,
        order,
        { req, res },
        emit,
      );
      // Places the order
      return isSuccessful && handlePlaceOrder(order, fraudDetectionStatus);
    }
    return false;
  };

  const saveAddresses = ({ currentCustomer }, order) => {
    if (currentCustomer.addressBook) {
      const allAddresses = addressHelpers.gatherShippingAddresses(order);
      allAddresses.forEach((address) => {
        if (
          !addressHelpers.checkIfAddressStored(
            address,
            currentCustomer.addressBook.addresses,
          )
        ) {
          addressHelpers.saveAddress(
            address,
            currentCustomer,
            addressHelpers.generateAddressName(address),
          );
        }
      });
    }
  };

  const isAdyen = hasAdyenPaymentMethod(currentBasket);

  if (!isAdyen) {
    return next();
  }

  const isValidTransaction = handleTransaction(
    currentBasket,
    { res, req },
    emit,
  );
  if (isValidTransaction) {
    const order = COHelpers.createOrder(currentBasket);

    saveAddresses(req, order);

    const isOrderCreated = handleCreateOrder(order);
    if (isOrderCreated) {
      COHelpers.sendConfirmationEmail(order, req.locale.id);

      // Reset usingMultiShip after successful Order placement
      req.session.privacyCache.set('usingMultiShipping', false);

      res.json({
        error: false,
        orderID: order.orderNo,
        orderToken: order.orderToken,
        continueUrl: URLUtils.url('Order-Confirm').toString(),
      });
      return emit('route:Complete');
    }
  }
  return undefined;
}

module.exports = createOrder;
