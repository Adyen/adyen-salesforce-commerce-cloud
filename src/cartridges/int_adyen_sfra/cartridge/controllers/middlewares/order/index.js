const confirm = require('./confirm');

module.exports = {
  confirm,
};
