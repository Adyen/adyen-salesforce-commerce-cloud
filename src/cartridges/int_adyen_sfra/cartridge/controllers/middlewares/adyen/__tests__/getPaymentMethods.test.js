/* eslint-disable global-require */
import {isAvailable} from "../../../../../../../../jest/__mocks__/dw/order/BasketMgr";

let getPaymentMethods;
let PaymentMgr;
let adyenHelper;
let BasketMgr;
let CustomerMgr;
let req;
let res;
beforeEach(() => {
  const { adyen } = require('../../index');
  getPaymentMethods = adyen.getPaymentMethods;
  PaymentMgr = require('dw/order/PaymentMgr');
  adyenHelper = require('*/cartridge/scripts/util/adyenHelper');
  BasketMgr = require('dw/order/BasketMgr');
  jest.clearAllMocks();
  res = { json: jest.fn() };
  req = {
    locale: { id: 'NL' },
    currentCustomer: { profile: { customerNo: 'mocked_customerNo' } },
  };
});

describe('Get Payment Methods', () => {
  it('should get country code', () => {
    getPaymentMethods(req, res, jest.fn());
    expect(BasketMgr.getCountryCode).toHaveBeenCalledTimes(1);
  });
  it('should get customer by customer number', () => {
    getPaymentMethods(req, res, jest.fn());
    expect(adyenHelper.getCustomer).toHaveBeenCalledWith(
      req.currentCustomer,
    );
  });
  it('should call get terminals if isActive', () => {
    getPaymentMethods(req, res, jest.fn());
    expect(PaymentMgr.isActive).toBeCalledTimes(1);
  });
  it('should handle installments when basket has total', () => {
    getPaymentMethods(req, res, jest.fn());
    expect(adyenHelper.getCurrencyValueForApi).toBeCalledTimes(1);
    expect(res.json).toMatchSnapshot();
  });
  it('should handle installments when basket has no total', () => {
    BasketMgr.isAvailable.mockImplementation(() => false);
    getPaymentMethods(req, res, jest.fn());
    expect(adyenHelper.getCurrencyValueForApi).toBeCalledTimes(0);
    expect(res.json).toMatchSnapshot();
  });
  it('should return response without installments', () => {
    adyenHelper.getCreditCardInstallments.mockImplementation(() => false);
    getPaymentMethods(req, res, jest.fn());
    expect(res.json).toMatchSnapshot();
  });
  it('should return response without connected terminals', () => {
    PaymentMgr.isActive.mockImplementation(() => false);
    getPaymentMethods(req, res, jest.fn());
    expect(res.json).toMatchSnapshot();
  });
});
