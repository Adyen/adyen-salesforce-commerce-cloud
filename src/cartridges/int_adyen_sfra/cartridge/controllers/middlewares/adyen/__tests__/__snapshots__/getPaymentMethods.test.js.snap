// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get Payment Methods should handle installments when basket has no total 1`] = `[MockFunction]`;

exports[`Get Payment Methods should handle installments when basket has total 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "AdyenConnectedTerminals": Object {
          "foo": "bar",
        },
        "AdyenDescriptions": Array [
          Object {
            "brandCode": "visa",
            "description": "mocked_hpp.description.visa",
          },
        ],
        "AdyenPaymentMethods": Object {
          "paymentMethods": Array [
            Object {
              "type": "visa",
            },
          ],
        },
        "ImagePath": "mocked_loading_contextimages/logos/medium/",
        "amount": Object {
          "currency": "EUR",
          "value": 1000,
        },
        "countryCode": "NL",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Get Payment Methods should return response without connected terminals 1`] = `[MockFunction]`;

exports[`Get Payment Methods should return response without installments 1`] = `[MockFunction]`;
